dali/core/mm/default_resources.cc

```diff
@@ -0,0 +1,282 @@
+// Copyright (c) 2021, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <stdexcept>
+#include <cstring>
+#include "dali/core/mm/default_resources.h"
+#include "dali/core/error_handling.h"
+#include "dali/core/format.h"
+#include "dali/core/mm/malloc_resource.h"
+#include "dali/core/device_guard.h"
+#include "dali/core/mm/async_pool.h"
+#include "dali/core/mm/composite_resource.h"
+#include "rmm/mr/device/owning_wrapper.hpp"
+#include "rmm/mr/device/cuda_memory_resource.hpp"
+#include "rmm/mr/device/pool_memory_resource.hpp"
+
+namespace dali {
+namespace mm {
+
+namespace {
+
+template <typename T>
+inline std::shared_ptr<T> wrap(T *p, bool own) {
+  if (own)
+    return std::shared_ptr<T>(p);
+  else
+    return std::shared_ptr<T>(p, [](T*){});
+}
+
+struct DefaultResources {
+  ~DefaultResources() {
+    ReleasePinned();
+    ReleaseDevice();
+    ReleaseHost();
+  }
+
+  static DefaultResources &instance() {
+    static DefaultResources dr;
+    return dr;
+  }
+
+  std::shared_ptr<host_memory_resource> host;
+  std::shared_ptr<pinned_async_resource> pinned_async;
+  std::vector<std::shared_ptr<device_async_resource>> device;
+
+  void ReleasePinned() {
+    Release(pinned_async);
+  }
+
+  void ReleaseDevice() {
+    Release(device);
+  }
+
+  void ReleaseHost() {
+    Release(host);
+  }
+
+ private:
+  template <typename Ptr>
+  void Release(Ptr &p) noexcept(false) {
+    if (cudaGetLastError() == cudaErrorCudartUnloading) {
+      abandon(p);
+    } else {
+      p = {};
+    }
+  }
+
+  template <typename T>
+  void abandon(std::vector<T> &v) {
```

Comment:
```suggestion
  void Abandon(std::vector<T> &v) {
```
nitpick: Capitalize to be consistent with other functions in this struct